plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobContainerRetriesTests.java

```diff
@@ -215,11 +234,49 @@ public void testReadBlobWithRetries() throws Exception {
         }
     }
 
+    public void testReadBlobInSmallChunks() throws Exception {
+        final int maxRetries = randomIntBetween(1, 5);
+        final int chunkSize = randomIntBetween(4096, 4096 * 2);
+        final byte[] bytes = randomBlobContent();
+        httpServer.createContext("/account/container/read_blob_small_chunks", exchange -> {
+            try {
+                Streams.readFully(exchange.getRequestBody());
+                if ("HEAD".equals(exchange.getRequestMethod())) {
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.getResponseHeaders().add("x-ms-blob-content-length", String.valueOf(bytes.length));
+                    exchange.getResponseHeaders().add("Content-Length", String.valueOf(bytes.length));
+                    exchange.getResponseHeaders().add("x-ms-blob-type", "blockblob");
+                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), -1);
+                } else if ("GET".equals(exchange.getRequestMethod())) {
+                    final int rangeStart = getRangeStart(exchange);
+                    assertThat(rangeStart, lessThan(bytes.length));
+                    int rangeSize = Math.min(chunkSize, bytes.length - rangeStart);
+                    exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+                    exchange.getResponseHeaders().add("Content-Range",
+                        "bytes " + rangeStart + "-" + (rangeStart + rangeSize) + "/" + bytes.length);
+                    exchange.getResponseHeaders().add("x-ms-blob-content-length", String.valueOf(rangeSize));
+                    exchange.getResponseHeaders().add("Content-Length", String.valueOf(rangeSize));
+                    exchange.getResponseHeaders().add("x-ms-blob-type", "blockblob");
+                    exchange.getResponseHeaders().add("ETag", UUIDs.base64UUID());
+                    exchange.sendResponseHeaders(RestStatus.PARTIAL_CONTENT.getStatus(), rangeSize);
+                    exchange.getResponseBody().write(bytes, rangeStart, rangeSize);
+                }
+            } finally {
+                exchange.close();
+            }
+        });
+
+        final BlobContainer blobContainer = createBlobContainer(maxRetries);
+        try (InputStream inputStream = blobContainer.readBlob("read_blob_small_chunks")) {
+            assertArrayEquals(bytes, BytesReference.toBytes(Streams.readFully(inputStream)));
+        }
+    }
+
     public void testReadRangeBlobWithRetries() throws Exception {
```

Comment:
This test fails right now, since the SDK triggers a `HEAD` request to get the blob size. The problem comes when a range that goes beyond the blob end the API would return an error `416 (Requested Range Not Satisfiable)`. My guess is that the expectation is that client would know the blob length beforehand and won't request past that boundary, is that correct?

Reply:
Additionally, we could be a little bit more efficient and avoid a copy + allocation per chunk read request if we implement it ourselves.

Reply:
As discussed, let's do it and try to avoid the extra HEAD + copying if we can.